### DISCUSSION TITLE
@Embedded와 @Embeddable은 각각 무슨 역할을 하나요?

### DISCUSSION THREAD

**Discussion by Kimprodp** [2024-woowacourse-study/level-interview#166] [Spring Data]
.

**Comment by Kimprodp** [comment]
두 어노테이션 모두 Entity 객체에서 값 타입을 정의하는 곳에 사용되며, @embeddable는 값 타입의 클래스에 사용되고, @Embedded는 값 타입을 사용하는 곳에 사용한다. 해당 어노테이션을 사용하면 값 타입 클래스의 필드들이 엔티티 클래스의 테이블에 컬럼으로 포함된다.

해당 어노테이션으로 정의된 값 객체는 테이블에 매핑될 때, 값 객체의 필드들을 모두 등록한다. JPA의 Entity 처럼 해당 값 객체도 기본 생성자를 필요로 한다.

**Comment by seunghye218** [comment]
`@Embedded`와 `@Embeddable`는 JPA(Java Persistence API)에서 사용되는 애노테이션으로, 엔티티의 속성을 더 잘 구조화하고 재사용성을 높이기 위해 사용됩니다. 이 두 애노테이션은 함께 사용되며, 각기 다른 역할을 수행합니다.

### @Embeddable

`@Embeddable`은 JPA 엔티티 클래스에서 사용되는, 다른 엔티티에 내장될 수 있는(embeddable) 클래스를 정의할 때 사용됩니다. 즉, 이 애노테이션은 해당 클래스가 다른 엔티티의 속성으로 포함될 수 있음을 나타냅니다.
- 특징
  - 내장될 수 있는 클래스임을 나타냅니다.
  - 데이터베이스 테이블에 매핑될 때, 독립된 테이블이 아닌 다른 엔티티의 테이블에 포함됩니다.
  - 보통 재사용 가능한 값 타입 객체(Value Object)를 정의할 때 사용됩니다.

### @Embedded

`@Embedded`는 JPA 엔티티 클래스 내에서 `@Embeddable`로 정의된 클래스의 인스턴스를 내장된 속성으로 사용할 때 사용됩니다. 즉, 엔티티 클래스에 내장된 클래스의 속성을 포함시키기 위해 사용됩니다.

- 특징
  - `@Embeddable`로 정의된 클래스의 인스턴스를 엔티티의 속성으로 포함시킵니다.
  - 내장된 클래스의 모든 필드는 상위 엔티티의 데이터베이스 테이블의 열로 매핑됩니다.
